### PR TITLE
add basic wasm decoder

### PR DESCRIPTION
**Pull request by MierenManz**

### CODE REVIEW

File: wasm/src/decoder/common.rs
Comment (Phosra) [inline wasm/src/decoder/common.rs:66]: This is plural, how about `results`?

Comment (Phosra) [inline wasm/src/decoder/common.rs:66]: Maybe these should be `Box<[ValueKind]>` to not indicate that the lengths will be mutated?

**Review by Phosra — COMMENTED** [review]
I also don't think using plain u8 works for deciding, because Wasm is a variable-sized instruction set.

File: wasm/src/decoder/code_section.rs
Comment (Phosra) [inline wasm/src/decoder/code_section.rs:7]: Nit: I believe there's an alias to 0, as Start, or something of the sort

File: wasm/src/decoder/code_section.rs
Comment (Phosra) [inline wasm/src/decoder/code_section.rs:17]: Hmm, this could extend via an interior?
`locals.extend((0..x.0).map(|_| x.1)` or maybe `locals.resize(locals.len() + x.0, x.1)` or maybe this is overkill anyways

File: wasm/src/decoder/mod.rs
Comment (Phosra) [inline wasm/src/decoder/mod.rs:80]: What does `unsigned` yield here? If it reads larger than a u32, too many bytes may be read, and the upper bytes are truncated..

File: wasm/src/decoder/type_section.rs
Comment (Phosra) [inline wasm/src/decoder/type_section.rs:8]: Does this allow modules that lack functions?
(I don't know how this part is being used)

File: wasm/src/error.rs
Comment (Phosra) [inline wasm/src/error.rs:10]: On its own, I don't think it's obvious as to what the u8 represents here?
As in, it's not clear as to whether it was the expected amount, or how many were read, even though I'd expect the latter to be usize.

File: wasm/src/lib.rs
Comment (Phosra) [inline wasm/src/lib.rs:6]: Maybe this should use `todo!()`, as it doesn't seem complete?

**Review by Phosra — COMMENTED** [review]
I'd suggest more `todo!()` instead of commented out code, really

File: wasm/src/common.rs
Comment (Phosra) [reply wasm/src/common.rs]: I later saw the `FnType`, so I think this is a non-issue

File: wasm/src/decoder/type_section.rs
Comment (MierenManz) [reply wasm/src/decoder/type_section.rs:8]: yes. This only get's called when modules do have a type_section and thus a function saction

File: wasm/src/error.rs
Comment (MierenManz) [reply wasm/src/error.rs:10]: Should I add a comment? something like `// TooManyBytes(expected)`

File: wasm/src/decoder/mod.rs
Comment (MierenManz) [reply wasm/src/decoder/mod.rs:80]: Gonna change this to `decode_u32` and not `read::unsigned`

File: wasm/src/decoder/code_section.rs
Comment (MierenManz) [reply wasm/src/decoder/code_section.rs:17]: Was trying that. But I wasn't able to figure it out. Gonna change that

File: wasm/src/decoder/code_section.rs
Comment (MierenManz) [reply wasm/src/decoder/code_section.rs:7]: `SeekFrom::Start` refers to the first byte and not the current one. And I do need the current one

**Comment by MierenManz** [comment]
>I also don't think using plain u8 works for deciding, because Wasm is a variable-sized instruction set.

So change the instructions from `Vec<u8>` to `Vec<Instruction>` where `Instruction` is a enum of all possible instructions?

File: wasm/src/decoder/code_section.rs
Comment (Phosra) [reply wasm/src/decoder/code_section.rs:17]: I'll do it in a follow-up PR

File: wasm/src/error.rs
Comment (Phosra) [reply wasm/src/error.rs:10]: Could be `TooManyBytes { expected_bytes: u8 }`

Comment (MierenManz) [reply wasm/src/error.rs:10]: Sgtm

File: wasm/src/decoder/code_section.rs
Comment (MierenManz) [reply wasm/src/decoder/code_section.rs:17]: You should be able to just push to this pr branch

File: wasm/src/decoder/mod.rs
Comment (Phosra) [inline wasm/src/decoder/mod.rs:78]: This is an example where I'd suggest `todo!(...)/unimplemented!(...)`

File: wasm/src/decoder/mod.rs
Comment (Phosra) [inline wasm/src/decoder/mod.rs:88]: Here too

File: wasm/src/decoder/mod.rs
Comment (Phosra) [inline wasm/src/decoder/mod.rs:93]: Ditto

**Review by Phosra — COMMENTED** [review]
> > I'd suggest more todo!() instead of commented out code, really
> 
> There is a `unimplemented!()` somewhere

**Comment by Phosra** [comment]
> > I also don't think using plain u8 works for deciding, because Wasm is a variable-sized instruction set.
> 
> So change the instructions from `Vec<u8>` to `Vec<Instruction>` where `Instruction` is a enum of all possible instructions?

Your call, but how are you planning on doing it otherwise with just `u8`?

**Comment by MierenManz** [comment]
> Your call, but how are you planning on doing it otherwise with just u8?

Essentially have a transformer from a Vec<u8> directly to cranelift ir rather than to Instruction and from Instruction to ir

**Comment by Phosra** [comment]
> > Your call, but how are you planning on doing it otherwise with just u8?
> 
> Essentially have a transformer from a Vec<u8> directly to cranelift ir rather than to Instruction and from Instruction to ir

I'm specifically talking about instructions that exceed the size of u8? And we might want it to be more general, because we may decide to eventually drop CraneLift

File: wasm/src/decoder/code_section.rs
Comment (Phosra) [reply wasm/src/decoder/code_section.rs:17]: I'll save it for an optimisation PR when we have tests and benchmarking.

**Comment by MierenManz** [comment]
Those would have additional bytes being read. The transformer wouldn't take only a u8 and convert it to a Instruction but rather take a `std::io::Read` or slice and consume the bytes necessary there and then convert that

File: wasm/src/decoder/code_section.rs
Comment (Phosra) [reply wasm/src/decoder/code_section.rs:7]: Yup, I forgot the API

File: wasm/src/decoder/common.rs
Comment (MierenManz) [reply wasm/src/decoder/common.rs:67]: your call

Comment (MierenManz) [reply wasm/src/decoder/common.rs:67]: sgtm

**Comment by MierenManz** [comment]
After this has landed I'm gonna remove the leb128 dep and just write a quick macro for decoding varint etc. Because right now it may return an error `TooManyBytes { expected: 10, found: 11 }` eventho you decode a u32

File: wasm/src/common.rs
Comment (MierenManz) [reply wasm/src/common.rs]: Will add those later

File: wasm/src/common.rs
Comment (MierenManz) [reply wasm/src/common.rs]: would a new `Error` variant word. `UnknownExternalKind` and for `ValueKind` the `UnknownValueKind`?

File: wasm/src/common.rs
Comment (Phosra) [reply wasm/src/common.rs]: We could have their decoding early, and just their usage as  `unimplemented!()`.

File: wasm/src/decoder/common.rs
Comment (Phosra) [reply wasm/src/decoder/common.rs]: In .wat, it's singular "param" though too `(param i32 i32 i32)`, and arguably we should just spell it out for "parameters." I'd say they should both be plural.

File: wasm/src/decoder/util.rs
Comment (Phosra) [inline wasm/src/decoder/util.rs:20]: Wouldn't this iterator be empty?

File: wasm/src/decoder/util.rs
Comment (Phosra) [inline wasm/src/decoder/util.rs:39]: I wouldn't expect this to even move the cursor, wouldn't you want to save the last position?

File: wasm/src/error.rs
Comment (Phosra) [inline wasm/src/error.rs:26]: Could avoid the allocation
```suggestion
            Self::TooManyBytes { expected, found } => write!(f, "Varint is too long! Expected {expected} bytes but found {found}")?,
```

File: wasm/src/error.rs
Comment (Phosra) [inline wasm/src/error.rs:21]: every path ends in `?`, could that be moved to the end of the `match`?

File: wasm/src/decoder/util.rs
Comment (MierenManz) [reply wasm/src/decoder/util.rs:20]: clippy didn't like my previous solution and recommended this to me instead.

previous solution was below
```rs
let mut v = Vec::with_capacity(length);

for i in 0..length {
    v[i] = func(reader)?;
}
```

File: wasm/src/decoder/util.rs
Comment (MierenManz) [reply wasm/src/decoder/util.rs:39]: well we would need to determine how many bytes were read. and by subtracting the begin position from the end we should get the difference. Eg how many bytes were read than expected

File: wasm/src/decoder/util.rs
Comment (Phosra) [reply wasm/src/decoder/util.rs:20]: It's because clippy lacks the knowledge that the Vec's *length is zero*.

This is just a bug, you want to push upon each iteration, not write into it via `*x =` or even `v[i] = `.

Comment (Phosra) [reply wasm/src/decoder/util.rs:20]:
```rust
for _ in 0..length {
 v.push(func(reader)?);
}
```

Comment (MierenManz) [reply wasm/src/decoder/util.rs:20]: oh yeah I see. Stupid mistake from myself. Weirdly enough I did do it correctly in the code section

```rs
for _ in 0..x.0 {
            locals.push(x.1);
        }
        ```

**Review by Phosra — APPROVED** [review]
I'm still doubtful about the seeking code, but I'll double-check that with some others later.

**Comment by MierenManz** [comment]
https://doc.rust-lang.org/std/io/trait.Seek.html#method.stream_position Check out this method. It's an alias for the seek

**Comment by Phosra** [comment]
Ah, I hadn't checked the return value of seek, that makes sense. LGTM

**Pull request merged**
